File: api/Features/Videos/Video.cs
namespace api.Features.Videos;

public enum VideoStatus { Processing, Ready, Failed }

public class Video
{
    private Video() { }

    public Video(string title, string description, Guid uploadedBy)
    {
        Title = title;
        Description = description;
        FilePath = string.Empty;
        UploadedBy = uploadedBy;
        CreatedAt = DateTimeOffset.UtcNow;
        Status = VideoStatus.Processing;
    }

    public Guid Id { get; private set; }
    public string Title { get; private set; } = null!;
    public string Description { get; private set; } = null!;
    public List<Tag> Tags { get; private set; } = [];
    public string FilePath { get; private set; } = null!;
    public Guid UploadedBy { get; private set; }
    public DateTimeOffset CreatedAt { get; private set; }
    public VideoStatus Status { get; private set; }

    public void SetReady(string filePath)
    {
        FilePath = filePath;
        Status = VideoStatus.Ready;
    }

    public void SetFailed()
    {
        Status = VideoStatus.Failed;
    }

    public void Update(string title, string description, List<Tag> tags)
    {
        Title = title;
        Description = description;
        Tags.Clear();
        Tags.AddRange(tags);
    }
}

File: client/lib/models/video.dart
import 'package:client/models/user.dart';

enum VideoStatus { processing, ready, failed }

class Video {
  final String id;
  final String title;
  final String description;
  final List<String> tags;
  final User uploader;
  final DateTime createdAt;
  final VideoStatus status;
  final int likeCount;
  final bool isLiked;

  Video({
    required this.id,
    required this.title,
    required this.description,
    required this.tags,
    required this.uploader,
    required this.createdAt,
    required this.status,
    required this.likeCount,
    required this.isLiked,
  });

  factory Video.fromJson(Map<String, dynamic> json) => Video(
    id: json['id'],
    title: json['title'],
    description: json['description'],
    tags: List<String>.from(json['tags']),
    uploader: User.fromJson(json['uploader']),
    createdAt: DateTime.parse(json['createdAt']),
    status: _parseStatus(json['status']),
    likeCount: json['likeCount'] as int,
    isLiked: json['isLikedByMe'] as bool,
  );

  static VideoStatus _parseStatus(dynamic s) {
    switch ((s as String?)?.toLowerCase()) {
      case 'ready':
        return VideoStatus.ready;
      case 'failed':
        return VideoStatus.failed;
      default:
        return VideoStatus.processing;
    }
  }
}
